### PR TITLE
[IR] Add overflow check in AllocaInst::getAllocationSize

### LINKED ISSUE #91380: AllocaInst::getAllocationSize/AllocaInst::getAllocationSizeInBits should check for overflow
Currently, the code just blindly multiplies the size, and can therefore return a nonsense result.  (The behavior isn't really defined in the overflow case, but reasoning based on the overflowed computation is likely to lead to weird results.)

https://github.com/llvm/llvm-project/blob/7115ed0fff027b65fa76fdfae215ed1382ed1473/llvm/lib/Transforms/Utils/InlineFunction.cpp#L2622 has code to compute the allocation size while conservatively checking for overflow; we should use similar logic, then make the inliner call getAllocationSize().  Maybe use the multiply-with-overflow helper from CheckedArithmetic.h instead of writing out that code.

### PR DESCRIPTION
**Pull request by tszhin-swe**
Fixes #91380

### CODE REVIEW

File: llvm/lib/IR/Instructions.cpp [hunk 1]
```diff
@@ -60,22 +61,33 @@ static cl::opt<bool> DisableI2pP2iOpt(
 std::optional<TypeSize>
 AllocaInst::getAllocationSize(const DataLayout &DL) const {
   TypeSize Size = DL.getTypeAllocSize(getAllocatedType());
-  if (isArrayAllocation()) {
-    auto *C = dyn_cast<ConstantInt>(getArraySize());
-    if (!C)
-      return std::nullopt;
-    assert(!Size.isScalable() && "Array elements cannot have a scalable size");
-    Size *= C->getZExtValue();
+  if (!isArrayAllocation()) {
```
Comment (tschuett) [inline llvm/lib/IR/Instructions.cpp:64]: Why do you need to invert the condition?

File: llvm/lib/IR/Instructions.cpp [hunk 2]
```diff
@@ -60,22 +61,33 @@ static cl::opt<bool> DisableI2pP2iOpt(
...
+    return Size;
   }
-  return Size;
+  auto *C = dyn_cast<ConstantInt>(getArraySize());
+  if (!C)
+    return std::nullopt;
+  assert(!Size.isScalable() && "Array elements cannot have a scalable size");
+  auto checkedProd =
+      checkedMulUnsigned(Size.getKnownMinValue(), C->getZExtValue());
+  if (!checkedProd) {
```
Comment (tschuett) [inline llvm/lib/IR/Instructions.cpp:73]: Please remove redundant braces.

File: llvm/lib/IR/Instructions.cpp (hunk 1 above)
Comment (tszhin-swe) [reply llvm/lib/IR/Instructions.cpp:64]: early exit for better readability

Comment (tschuett) [reply llvm/lib/IR/Instructions.cpp:64]: I would prefer to keep the existing code style and only fix the overflow issue.

Comment (tszhin-swe) [reply llvm/lib/IR/Instructions.cpp:64]: Sure, it's my first time contributing to llvm, so I was just trying to conform to the coding standard which says to use early exit to minimize nested if's.

File: llvm/lib/IR/Instructions.cpp (hunk 2 above)
Comment (tszhin-swe) [reply llvm/lib/IR/Instructions.cpp]: Fixed

File: llvm/lib/IR/Instructions.cpp (hunk 1 above)
Comment (tschuett) [reply llvm/lib/IR/Instructions.cpp]: The diff becomes smaller and it is easier to review when you focus on the overflow check.

File: llvm/lib/IR/Instructions.cpp [hunk 3]
```diff
@@ -60,22 +61,29 @@ static cl::opt<bool> DisableI2pP2iOpt(
...
 
 std::optional<TypeSize>
 AllocaInst::getAllocationSizeInBits(const DataLayout &DL) const {
   std::optional<TypeSize> Size = getAllocationSize(DL);
-  if (Size)
-    return *Size * 8;
-  return std::nullopt;
+  if (!Size)
+    return std::nullopt;
+  auto CheckedProd = checkedMulUnsigned((*Size).getKnownMinValue(),
```
Comment (nikic) [inline llvm/lib/IR/Instructions.cpp:82]:
```suggestion
  auto CheckedProd = checkedMulUnsigned(Size->getKnownMinValue(),
```

File: llvm/lib/IR/Instructions.cpp [hunk 4]
```diff
@@ -60,22 +61,29 @@ static cl::opt<bool> DisableI2pP2iOpt(
...
+                                        static_cast<TypeSize::ScalarTy>(8));
+  if (!CheckedProd)
+    return std::nullopt;
+  return TypeSize::get(*CheckedProd, (*Size).isScalable());
```
Comment (nikic) [inline llvm/lib/IR/Instructions.cpp:86]:
```suggestion
  return TypeSize::get(*CheckedProd, Size->isScalable());
```

File: llvm/lib/IR/Instructions.cpp [hunk 5]
```diff
@@ -60,22 +61,29 @@ static cl::opt<bool> DisableI2pP2iOpt(
...
-    Size *= C->getZExtValue();
-  }
-  return Size;
+  if (!isArrayAllocation())
+    return Size;
+  auto *C = dyn_cast<ConstantInt>(getArraySize());
+  if (!C)
+    return std::nullopt;
+  assert(!Size.isScalable() && "Array elements cannot have a scalable size");
+  auto checkedProd =
```
Comment (nikic) [inline llvm/lib/IR/Instructions.cpp:70]:
```suggestion
  auto CheckedProd =
```

File: llvm/lib/IR/Instructions.cpp (hunk 3 above)
Comment (tszhin-swe) [reply llvm/lib/IR/Instructions.cpp]: Fixed

File: llvm/lib/IR/Instructions.cpp (hunk 4 above)
Comment (tszhin-swe) [reply llvm/lib/IR/Instructions.cpp]: Fixed

File: llvm/lib/IR/Instructions.cpp (hunk 5 above)
Comment (tszhin-swe) [reply llvm/lib/IR/Instructions.cpp]: Fixed

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**